gitcoin: AlluoLPV2, new contract functions, fixes and a suggestion

pentatonictritones opened the pull request:
Hi,

According to the bounty, I have rewritten the deposit, withdraw, getBalance functions with two helper functions so that interest can be calculated more efficiently. I have also written tests in hardhat for this new version. I've tried to follow your conventions as much as possible and kept as much code in similar places so that all your events are triggered.

Because I haven't received clarification about what exactly you mean, I have omitted the exchange part - but the implementation should be very straight forward (couple lines of code anyways).

While I can write you complete documenation for all pieces of code added, I'll just explain the core concepts here:

1.)  The new Epoch array 

Essentially, your previous implementations + the other pull request is not an accurate way to calculate interest over time. If the interest rates change over time (admin setting them), then the user should receive the appropriate interest for depositing in that period and then get an adjusted rate from that point onwards. However, the current implementations just wipe out the historical interest rate information.

Whilst you can create a workaround by updating everyone's DF whenever interest is updated, that is very inefficient. Therefore, it is better to keep track of interest rate regime changes through 'epochs' where a new epoch is created every time interest rates are changed.

2.) Two new functions _getCompoundedBalance() and _getPeriodsPerEpoch:

Therefore, _getCompoundedBalance() is a function that calculates how many compounding periods the user should get for each relevant epoch and then returns the correct balance.

For example:

- User deposits 100 USD on epoch 0, interest rate 8% APY,

- 20 days pass and interest rate is changed to 7.7% APY

- then 30 days pass and the interest rate is changed to 9.3% APY,

- then 10 more days pass (no rate changes).

- This calculation does:   20 days at 8% APY, 30 days at 7.7% APY, 10 days at 9.3% APY.


*** Whilst this implementation you are going for (setting interest rates manually, then giving everyone allowances to withdraw their compounded balance) is easy to set up, **I would seriously consider switching to a model where user balances are rebased directly and interest rates can be calculated from the size of the rebase.** 

You can see how olympus DAO does it for reference. This is because it removes the possibility of bank runs (see macrofinance suspension of convertibility) and is more permissionless. So how it would work is, your multisig wallet dumps investment returns into the LP and then triggers a rebase based on the profits, not off an interest rate.

Happy to work on any reasonable iterations or future projects.
PT


Once I get feedback on this, I'll actually do the implementation for the other bounty you wanted. I think it is better for us to be on the same page before I move forward.

0xtuytuy commented:
@pentatonictritones thanks for the submission.

@GetAlluo/protocole-squad could someone help review the work here please 🚀

zayk00v commented:
Hi, thanks for your contribution. 
Are you sure about your statement that the current implementation calculates an incorrect APY if the percentage has been changed? Have you checked the APY calculation tests? there is at least one with a changing interest in the span of one year. 
The current implementation has one downside, which is that the user can get even a little more rewards, because basically what the updating DF  does is an autoclaim for everyone. 
So anyway we are going to change it in the next release with a new version of alluoLp. But if you think that our reward calculation system is wrong, offer us a scenario in which it is visible, we will appreciate this.
About your implementation. We thought about the approach with  epochs at the very beginning, but were moved from it because of  gas inefficiency. but in any case, we are moving away from the current idea, and what we wanted to see in this bounty is a complete change in the reward system. something like new algorithm that change ratio between alluoLp and stables. 
Example:  at the beginning user will get 100 alluoLp for 100 stables, but after a year he will  get only 100 alluoLp for 108 stables, if interest was 8 whole year. maybe it wasn't so clear from the explanation, sorry for that. 
As i am writing it i get that it is even simpler than i though and then current implementation. So now if you will try console.log (DF) in update function in tests (like calling it every day) you will see how it gradually go from 100 to 108+, basically it is price(ratio) of alluoLp to stables. you need to calculate how much alluoLp you need to mint in deposit and burn in withdraw based on this growing ratio. it is pretty simple. 
Two more things: first maybe we need change formula of updating DF so it will not go over 108 in one year and will grow linearly. Second you need prepare migration function so we can call it after upgrading, claim all rewards for users(probably we will input all users addresses manually), refresh DF and so on. And please also dont forget about tests. dont worry about exchange in tests or if you want you can see how we exchanging  it now using curve Aave pool in deploy branch

pentatonictritones commented:
> Hi, thanks for your contribution. Are you sure about your statement that the current implementation calculates an incorrect APY if the percentage has been changed? Have you checked the APY calculation tests? there is at least one with a changing interest in the span of one year. The current implementation has one downside, which is that the user can get even a little more rewards, because basically what the updating DF does is an autoclaim for everyone. So anyway we are going to change it in the next release with a new version of alluoLp. But if you think that our reward calculation system is wrong, offer us a scenario in which it is visible, we will appreciate this. About your implementation. We thought about the approach with epochs at the very beginning, but were moved from it because of gas inefficiency. but in any case, we are moving away from the current idea, and what we wanted to see in this bounty is a complete change in the reward system. something like new algorithm that change ratio between alluoLp and stables. Example: at the beginning user will get 100 alluoLp for 100 stables, but after a year he will get only 100 alluoLp for 108 stables, if interest was 8 whole year. maybe it wasn't so clear from the explanation, sorry for that. As i am writing it i get that it is even simpler than i though and then current implementation. So now if you will try console.log (DF) in update function in tests (like calling it every day) you will see how it gradually go from 100 to 108+, basically it is price(ratio) of alluoLp to stables. you need to calculate how much alluoLp you need to mint in deposit and burn in withdraw based on this growing ratio. it is pretty simple. Two more things: first maybe we need change formula of updating DF so it will not go over 108 in one year and will grow linearly. Second you need prepare migration function so we can call it after upgrading, claim all rewards for users(probably we will input all users addresses manually), refresh DF and so on. And please also dont forget about tests. dont worry about exchange in tests or if you want you can see how we exchanging it now using curve Aave pool in deploy branch

Hi, thanks for the response.

Read about Alluo's product and it's very cool. I've been explaining to fintech people that it's a low hanging fruit to do this with debit cards using anchor but I think there are regulatory challenges with it here in the UK. Is Alluo going to add debit card support? That would be amazing as an offramp. Real value there.

So about the APY calculation - you're correct. It was my mistake. I didn't see that DF was being updated upon setInterest(). I wrote a couple tests and it does indeed give the correct value!

I understand what you mean now with the new system. I think it is just better.

So I will work on the following tasks then: 

1. Revamp functions to use an Interest Index 
2. Migration function 
3. Tests

I'll push an update with tests in 2-3 days.

0xtuytuy commented:
> Read about Alluo's product and it's very cool. I've been explaining to fintech people that it's a low hanging fruit to do this with debit cards using anchor but I think there are regulatory challenges with it here in the UK. Is Alluo going to add debit card support? That would be amazing as an offramp. Real value there.

Thanks! Some of the core contributors are fintech ppl with experience building fully regulated banks from scratch. We are planning to add a card (see your public roadmap here: https://roadmap.alluo.com/) we are already in discussions with multiple card providers and directly with a few card issuers (without dropping too big names 😮)

The challenge indeed is regulatory, Visa or Mastercard are not going to deal directly with a DAO, you need some sort of regulated (KYC/AML, etc) entity that can deal with Visa. This is something we are already in process of setting up 👍 

> Revamp functions to use an Interest Index
Migration function
Tests

Looking forward to your push 💯

pentatonictritones commented:
> Please note that there are 2 people working on the same bounty
> 
> * [gitcoin: AlluoLPV2, new contract functions, fixes and a suggestion #22](https://github.com/GetAlluo/alluo-protocol/pull/22)
> * [feat: https://github.com/GetAlluo/alluo-protocol/issues/17 #21](https://github.com/GetAlluo/alluo-protocol/pull/21)
> 
> @HarryTgerman and @pentatonictritones it might be worth joining forces 🤲

Hi @zayk00v  @0xtuytuy , I completed the following tasks:

1. Rewrote so that we now use an interest index to calculate withdrawals, deposits, balances.. etc.

2. The index doesn't go over 8% (if interest is set at 8%). See the tests in AlluoLpV4.ts. 

3. Migration function has been written to just iterate through a manual list and mint tokens in the new contract - if I understand it correctly, getBalance of the old contract returns USD value, so the migration function is written so that you can migrate balances later in time to mint correct numbers of AlluoLp tokens, not just on the day of deploying the contract.

4. I haven't touched other functions that you have written and normal ERC20 functions so they are safe if you've tested them.

5.  In case of any confusion, I'm using daily compounding (as I believe is appropriate for this application) so the interest rate is the value 'x' in  x^365 = 1 + APY. 

Also, it would be good if you could check event BurnedForWithdraw 
My implementation has the number of alluo tokens to be burnt for withdraw. If you want the USD value, you can change adjustedAmount --> _amount.

I think this should complete the scope of the bounty task. Please let me know if there are any small concerns you have and I'll get on it.

zayk00v commented:
Hi, i think we will not close this bounty yet. Here some comments:
1) Did you test what will happen if users will call your updateInterestIndex() more then every day, i might be wrong but it seams like if it will happen compoundingPeriods will always be 0 so as final interest. 
2)  Daily compounding is not what we are looking for, if you will open our mobile app you will see how your balance grow constantly. We want to see same with new implantation, and at the same time have fixed interest. So again constant auto compounding(like every block or with some small fixed time period), but at the end it should be as mach as it possible to main interest.
3) About migration. We want to stay on same address, so we need to migrate balances and change logic. How we see it: we upgrading contract with new logic inside, and leave all required function for balances migration. Pausing contract and call migrate function. inside migrate function we claiming user rewards and resetting DF(or InterestIndex). Unpausing contract.

pentatonictritones commented:
> Hi, i think we will not close this bounty yet. Here some comments:
> 
> 1. Did you test what will happen if users will call your updateInterestIndex() more then every day, i might be wrong but it seams like if it will happen compoundingPeriods will always be 0 so as final interest.
> 2. Daily compounding is not what we are looking for, if you will open our mobile app you will see how your balance grow constantly. We want to see same with new implantation, and at the same time have fixed interest. So again constant auto compounding(like every block or with some small fixed time period), but at the end it should be as mach as it possible to main interest.
> 3. About migration. We want to stay on same address, so we need to migrate balances and change logic. How we see it: we upgrading contract with new logic inside, and leave all required function for balances migration. Pausing contract and call migrate function. inside migrate function we claiming user rewards and resetting DF(or InterestIndex). Unpausing contract.

Hi,

1.  I accidentally removed my if statement there for daily compounding while removing it in my view loop because it doesn't need it there.  My bad
2.  I can set the fixed time period to like 20 seconds (which has enough margin from timestamp manipulation). You can't set it too low as you know because people may attack it and raise interestIndex too high. However, I think 20 seconds is ok as long as your app animation just displays the second-by-second changes purely cosmetically. Also, you're ok with your users paying a little bit more gas for more often compounding right?
3. OK I thought you meant migrating contracts but just upgrading instead. Will do.

zayk00v commented:
1) Of course, no problem
2) Ideally, we want to have a system in which it doesn't really matter how often the user calls the update, but in the end he will get a fixed percentage. 
About gas, we don't worry too much about it, because in the mobile app we pay for the transaction out of our pocket, and it's a polygon, but it wouldn't be good if at the end it required from user to get through loop  with 100+ iterations, so you also need to think maybe there are a few different ways of calculating interest, as in our current implementation. If you want, I'll try to find some links on which we baised our DF.
3) Cool.

pentatonictritones commented:
> 1. Sure, no problem
> 2. Ideally, we want to have a system in which it doesn't really matter how often the user calls the update, but in the end he will get a fixed percentage.
>    About gas, we don't worry too much about it, because in the mobile app we pay for the transaction out of our pocket, and it's a polygon, but it wouldn't be good if at the end it required from user to get through loop  with 100+ iterations, so you also need to think maybe there are a few different ways of calculating interest, as in our current implementation. If you want, I'll try to find some links on which we baised our DF.
> 3. Cool.

@zayk00v @0xtuytuy 

Hi,

I saw that you've pushed a new update to the master with quite a lot of changes, so I ended up cloning the current version of the master branch and I rewrote my contracts based on them so that they are upgradable.

What I have done:

1. Allow continuous compounding: Getbalance gets the compounded balance correct to the second, while the actual updateInterestIndex() function can be called every 60 seconds (or how ever long you set the interval) so that it updates correctly down to the second.

I ended up using math contracts from MakerDao as their implementation was what you were looking for. I checked the licenses and it is fine.

2. About migration: 

I wrote a migrate() which just sets the new system's variables, so I imagine these steps to work:

1. Claim balances of all users before upgrading
2. Upgrade contract to V4
3. Call migrate() function and all is complete.

This is because at t=0, ERC20 balance is 1:1 with the USD/whatever stablecoin deposit.


3. I wrote tests for all these cases, including the previous concerns you had about APY.

The only thing is my migration test: Just because of how there are many components that you have added very recently - like the liquidity buffer.. etc -  I have made a little shortcut so that errors from not setting variables do not occur.

I think this step is quite simple anyways given the implementation so I don't think there will be much issue with this.

I believe that this should wrap the bounty up nicely ^^

pentatonictritones commented:
Hi, realised there were issues with depositing and withdrawing tokens with decimals that are not 18. Didn't know USDC had 6 decimals. So I just added a small conversion in withdraw and deposit.

Also wrote all the tests to check that they work. Keep in mind that when getBalance is called, need to divide by 10**18 for USD value.

alexanderem49 commented:
Drafted as work is still in progress. Please, open this PR as work will be completed and ready for review.

pentatonictritones commented:
> Drafted as work is still in progress. Please, open this PR as work will be completed and ready for review.

Hi, I have completed all the tasks set and think it should be good to be reviewed.  Could you explain what you mean?

0xtuytuy commented:
@pentatonictritones I was trying to test your branch, there seem to be a few conflicts with the current master. Could you please fix this so we can review & approve ?

```| ~/Documents/alluo-protocol @ 19:04:45 - (remituyaerts) 
| => git merge master 
Auto-merging test/AlluoLpUpgradable.ts
Removing tasks/tasks.ts
Removing scripts/deployAlluoLpMintable.ts
Auto-merging package.json
CONFLICT (content): Merge conflict in package.json
Auto-merging package-lock.json
CONFLICT (content): Merge conflict in package-lock.json
Auto-merging hardhat.config.ts
CONFLICT (content): Merge conflict in hardhat.config.ts
Auto-merging contracts/mock/old/AlluoLpUpgradableMintable.sol
CONFLICT (modify/delete): .env.example deleted in HEAD and modified in master. Version master of .env.example left in tree.
Automatic merge failed; fix conflicts and then commit the result.

pentatonictritones commented:
> @pentatonictritones I was trying to test your branch, there seem to be a few conflicts with the current master. Could you please fix this so we can review & approve it?
> 
> If you can as well, please join our discord, we can give feedback faster this way [Link](https://discord.gg/jNaQF6sMxf)
> 
> ```
> | => git merge master 
> Auto-merging test/AlluoLpUpgradable.ts
> Removing tasks/tasks.ts
> Removing scripts/deployAlluoLpMintable.ts
> Auto-merging package.json
> CONFLICT (content): Merge conflict in package.json
> Auto-merging package-lock.json
> CONFLICT (content): Merge conflict in package-lock.json
> Auto-merging hardhat.config.ts
> CONFLICT (content): Merge conflict in hardhat.config.ts
> Auto-merging contracts/mock/old/AlluoLpUpgradableMintable.sol
> CONFLICT (modify/delete): .env.example deleted in HEAD and modified in master. Version master of .env.example left in tree.
> Automatic merge failed; fix conflicts and then commit the result.
> ```

Hi, I thought I fixed them but it seems like I forgot to git add haha. 

I changed the imports for the contracts and just checked by using git pull that it should merge without conflicts now and everything compiles fine on my end.

I've joined the discord as well - id is the same as here : pentatonictritones.

0-xec commented on contracts/AlluoLpV4.sol:
```diff
@@ -0,0 +1,284 @@
+
+    /// @notice migrates by setting new interest variables.
+    /// @dev Only call this after all balances are claimed (before upgrade is initiated)
+    ///  So 1. Claim all balances --> 2. Upgrade the contract  3. Call migrate()
```
we can loop and claim for all addresses one by one just before the upgrade

0xtuytuy commented on contracts/AlluoLpV4.sol:
```diff
@@ -0,0 +1,284 @@
+        _burn(msg.sender, adjustedAmount);
+        IERC20Upgradeable(_targetToken).safeTransfer(msg.sender, _amount);
+        emit BurnedForWithdraw(msg.sender, adjustedAmount);
+    }
```
@zayk00v is taking care of this to move faster

pentatonictritones commented:
> There are a few things left to do:
> 
> 1. implement the liquidity buffer
> 2. clean up the files with the merge headers
> 3. explain a bit the migration process

Hi,

1. I know someone else is working on this implementation.

2.  Should be done. I checked that merging works and npx hardhat test works without issue.

**3. Migration process:**

1. First, claim all balances by running a loop through all addresses holding ERC20 tokens. You should be able to find these addresses by looking at the Deposited Event emitted historically. Once you call claim for everyone, proceed to step 2.
2. Upgrade the contract to the latest version like you did for the previous V2--> V3.
3. Call migrate() from the admin multisig wallet. 

Keep in mind that it sets the interest rate to 8% APY on migrate() so if you want to change it, just call setInterest with the new rate.

Also, keep in mind that interestRate is **interest per second** (uint 10**27)so whatever APY you want, just find  x in ==> x^31536000 = 1 + APY

So if it APY = 8%,

x ^ 31536000 = 1.08,  
x is approx 1.00000000244... (get higher precision by adding more decimal places)

Then, set interestRate = x * 10**27.

I approximated it to a higher degree of accuracy so: 
      interestRate = 100000000244041*10**13;

which is same as 1.00000000244041 * 10**27
